l1-contracts/contracts/bridgehub/Bridgehub.sol

```diff
@@ -601,68 +601,82 @@ contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2StepUpgradeable, Paus
     }
 
-    /// @dev IL1AssetHandler interface, used to receive a chain on the settlement layer.
-    /// @param _assetId the assetId of the chain's STM
-    /// @param _bridgehubMintData the data for the mint
     function bridgeMint(
         uint256, // originChainId
         bytes32 _assetId,
         bytes calldata _bridgehubMintData
     ) external payable override onlyAssetRouter returns (address l1Receiver) {
-        (uint256 _chainId, bytes memory _stmData, bytes memory _chainMintData) = abi.decode(
-            _bridgehubMintData,
-            (uint256, bytes, bytes)
-        );
+        BridgehubSTMAssetData memory bridgeData = abi.decode(_bridgehubMintData, (BridgehubSTMAssetData));
+
         address stm = stmAssetIdToAddress[_assetId];
         require(stm != address(0), "BH: assetInfo 2");
-        require(settlementLayer[_chainId] != block.chainid, "BH: already current SL");
+        require(settlementLayer[bridgeData.chainId] != block.chainid, "BH: already current SL");
 
-        settlementLayer[_chainId] = block.chainid;
-        stateTransitionManager[_chainId] = stm;
+        settlementLayer[bridgeData.chainId] = block.chainid;
+        stateTransitionManager[bridgeData.chainId] = stm;
         address hyperchain;
-        if (hyperchainMap.contains(_chainId)) {
-            hyperchain = hyperchainMap.get(_chainId);
+        if (hyperchainMap.contains(bridgeData.chainId)) {
+            hyperchain = hyperchainMap.get(bridgeData.chainId);
         } else {
-            hyperchain = IStateTransitionManager(stm).forwardedBridgeMint(_chainId, _stmData);
+            hyperchain = IStateTransitionManager(stm).forwardedBridgeMint(bridgeData.chainId, bridgeData.stmData);
         }
 
-        messageRoot.addNewChainIfNeeded(_chainId);
-        _registerNewHyperchain(_chainId, hyperchain);
-        IZkSyncHyperchain(hyperchain).forwardedBridgeMint(_chainMintData);
+        messageRoot.addNewChainIfNeeded(bridgeData.chainId);
+        _registerNewHyperchain(bridgeData.chainId, hyperchain);
+        IZkSyncHyperchain(hyperchain).forwardedBridgeMint(bridgeData.chainData);
         return address(0);
     }
 
     /// @dev IL1AssetHandler interface, used to undo a failed migration of a chain.
     /// @param _chainId the chainId of the chain
     /// @param _assetId the assetId of the chain's STM
-    /// @param _data the data for the recovery
+    /// @param _data the data for the recovery.
     function bridgeRecoverFailedTransfer(
         uint256 _chainId,
         bytes32 _assetId,
         address _depositSender,
         bytes calldata _data
-    ) external payable override onlyAssetRouter onlyL1 {}
+    ) external payable override onlyAssetRouter onlyL1 {
+        BridgehubSTMAssetData memory stmAssetData = abi.decode(_data, (BridgehubSTMAssetData));
+        HyperchainCommitment memory chainCommitment = abi.decode(stmAssetData.chainData, (HyperchainCommitment));
```

Comment:
same here, you should just pass the along the used `stmAssetData.chainData` and never parse it